code/game/objects/effects/portals.dm

```diff
@@ -1,69 +1,143 @@
+	var/turf/open/atmos_destination	//Atmos link destination
 
-/obj/effect/portal/Bumped(mob/M as mob|obj)
-	teleport(M)
+/obj/effect/portal/attackby(obj/item/weapon/W, mob/user, params)
+	if(user && Adjacent(user))
+		user.forceMove(get_turf(src))
+		teleport(user)
+
+/obj/effect/portal/make_frozen_visual()
+	return
+
+/obj/effect/portal/Crossed(atom/movable/AM, oldloc)
+	if(get_turf(oldloc) == get_turf(linked))
+		return ..()
+	if(!teleport(AM))
+		return ..()
 
 /obj/effect/portal/attack_tk(mob/user)
 	return
 
 /obj/effect/portal/attack_hand(mob/user)
 	if(Adjacent(user))
+		user.forceMove(get_turf(src))
 		teleport(user)
 
-/obj/effect/portal/attackby(obj/item/weapon/W, mob/user, params)
-	if(user && Adjacent(user))
-		teleport(user)
+/obj/effect/portal/Initialize(mapload, _creator, _lifespan = 300, obj/effect/portal/_linked = null, automatic_link = TRUE, hard_target_override = null, atmos_link_override = null)
+	. = ..()
+	GLOB.portals += src
+	if(!istype(_linked) && automatic_link)
+		return INITIALIZE_HINT_QDEL
+	if(_lifespan > 0)
+		QDEL_IN(src, _lifespan)
+	if(!isnull(atmos_link_override))
+		atmos_link = atmos_link_override
+	link_portal(_linked)
+	hardlinked = automatic_link
+	creator = _creator
 
-/obj/effect/portal/make_frozen_visual()
-	return
+/obj/effect/portal/proc/link_portal(obj/effect/portal/newlink)
+	linked = newlink
+	if(atmos_link)
+		unlink_atmos()
+		link_atmos()
 
-/obj/effect/portal/New(loc, turf/target, creator=null, lifespan=300)
-	..()
-	GLOB.portals += src
-	src.target = target
-	src.creator = creator
+/obj/effect/portal/proc/link_atmos()
+	if(atmos_source || atmos_destination)
+		unlink_atmos()
+	if(!isopenturf(get_turf(src)))
+		return FALSE
+	if(linked)
+		if(isopenturf(get_turf(linked)))
+			atmos_source = get_turf(src)
+			atmos_destination = get_turf(linked)
+	else if(hard_target)
+		if(isopenturf(hard_target))
+			atmos_source = get_turf(src)
+			atmos_destination = hard_target
+	else
+		return FALSE
+	if(!istype(atmos_source) || !istype(atmos_destination))
+		return FALSE
+	LAZYINITLIST(atmos_source.atmos_adjacent_turfs)
+	LAZYINITLIST(atmos_destination.atmos_adjacent_turfs)
+	if(atmos_source.atmos_adjacent_turfs[atmos_destination] || atmos_destination.atmos_adjacent_turfs[atmos_source])	//Already linked!
+		return FALSE
+	atmos_source.atmos_adjacent_turfs[atmos_destination] = TRUE
+	atmos_destination.atmos_adjacent_turfs[atmos_source] = TRUE
+	atmos_source.air_update_turf(FALSE)
+	atmos_destination.air_update_turf(FALSE)
 
-	var/area/A = get_area(target)
-	if(A && A.noteleport) // No point in persisting if the target is unreachable.
-		qdel(src)
-		return
-	if(lifespan > 0)
-		QDEL_IN(src, lifespan)
+/obj/effect/portal/proc/unlink_atmos()
+	if(istype(atmos_source))
+		LAZYREMOVE(atmos_source.atmos_adjacent_turfs, atmos_destination)
+		atmos_source = null
+	if(istype(atmos_destination))
+		LAZYREMOVE(atmos_destination.atmos_adjacent_turfs, atmos_source)
+		atmos_destination = null

```

Comment:
This proc doesnt check if the turfs are adjucent, you can unlink nearby turfs from one another like this and they will stay unlinked